### PR TITLE
[MOS-288] Add community build

### PR DESCRIPTION
**Pull request by paweljonskim**
Główne zmiany:
- Style.hpp
- FontManager.hpp
- FontManager.cpp
- products/BellHybrid/CMakeLists.txt
- image/BellHybrid/CMakeLists.txt

### CODE REVIEW

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:60]:
```suggestion
        for (const auto& font : fontFiles) {
```

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:65]:
```suggestion
    RawFont *FontManager::loadFont(Fonts font, const std::string& path)
```

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:164]: `std::fclose(fd)` missing?

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:151]:
```suggestion
        for (const auto &entry : fonts_map::mapJson) {
```

File: cmake/modules/DownloadAsset.cmake
Comment (swichu91) [inline cmake/modules/DownloadAsset.cmake:105]:
```suggestion
if (NOT TARGET json-proprietary-target)
    download_asset_json(json-proprietary-target
                        ${CMAKE_CURRENT_SOURCE_DIR}/assets/assets_proprietary.json
                        ${CMAKE_BINARY_DIR}/sysroot/sys/current/
                        MuditaOSAssets
                        ${MUDITA_CACHE_DIR}
                        )
download_asset_release_json(json-common-target
                            ${CMAKE_CURRENT_SOURCE_DIR}/assets/assets_common.json
                            ${CMAKE_BINARY_DIR}/sysroot/sys/current/
                            MuditaOSPublicAssets
                            0.0.5
                            ${MUDITA_CACHE_DIR}
                            )
download_asset_release_json(json-community-target
                                ${CMAKE_CURRENT_SOURCE_DIR}/assets/assets_community.json
                                ${CMAKE_BINARY_DIR}/sysroot/sys/current/
                                MuditaOSPublicAssets
                                0.0.5
                                ${MUDITA_CACHE_DIR}
                                )
download_asset_release_json(json-rt1051-target
                            ${CMAKE_CURRENT_SOURCE_DIR}/assets/assets_rt1051.json
                            ${CMAKE_BINARY_DIR}/sysroot/sys/current/
                            MuditaOSPublicAssets
                            0.0.5
                            ${MUDITA_CACHE_DIR}
                            )
endif()
```

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:77]: Replace it with `std::unique_ptr<char[]> p(new char[fileSize])` or use plain std::vector. Then you won't need to manually manage memory.

**Review by swichu91 — COMMENTED** [review]
Please fix some minor issues, apart from that it looks good to me.

File: module-bsp/board/linux/hal/battery_charger/BatteryCharger.cpp
Comment (swichu91) [inline module-bsp/board/linux/hal/battery_charger/BatteryCharger.cpp:47]: Should it be changed?

Comment (paweljonskim) [reply module-bsp/board/linux/hal/battery_charger/BatteryCharger.cpp:47]: No! thanks for catching it

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:93]: What if `RawFont` returned nullptr?

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:119]: File descriptor ought to be closed if not used anymore

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:139]:
```suggestion
        const auto fontmapObjects = fontmapJson.object_items();
        const auto infoJson       = fontmapObjects["info"];
```

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:143]:
```suggestion
        const auto styleJson = fontmapObjects["style"];
```

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:219]:
```suggestion
        if (font == nullptr && not fonts.empty()) {
```

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:76]: There is no need for rewinding.

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:81]: Return value check missing.

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:75]:
```suggestion
        const auto fileSize = std::filesystem::file_size(path);
```

File: module-gui/gui/core/FontManager.cpp
Comment (swichu91) [inline module-gui/gui/core/FontManager.cpp:83]:
```suggestion
        const auto bytesRead = input.gcount();
```

File: module-gui/gui/widgets/NavBar.cpp
Comment (mjan-dev) [inline module-gui/gui/widgets/NavBar.cpp:46]: navbar medium should be huge?

File: products/BellHybrid/apps/application-bell-onboarding/windows/OnBoardingOnOffWindow.cpp
Comment (mjan-dev) [inline products/BellHybrid/apps/application-bell-onboarding/windows/OnBoardingOnOffWindow.cpp:31]: Why the change in size?

File: test/assets/assets_common.json
Comment (mjan-dev) [inline test/assets/assets_common.json:2]: what is the reason for this empty file?

**Review by mjan-dev — APPROVED** [review]
Approved with minor comments. Please respond before merging.

File: configure.sh
Comment (pholat) [inline configure.sh:77]: not needed, please remove - it's not like we add all options to configure & catalog directory.
Also - there is no need to - reconfiguration should be lightweight for this example.

File: image/CMakeLists.txt
Comment (pholat) [inline image/CMakeLists.txt:14]: please rename to `ASSETS_TYPE` for example, just `ASSETS` seems to ambiguous

File: image/CMakeLists.txt
Comment (pholat) [inline image/CMakeLists.txt:16]: why do we need a flag for RT1051 nowhere? 
@swichu91 won't it be better to just include rt1051 in `only rt1051` cmakes?

File: module-apps/apps-common/popups/BluetoothAuthenticatePopup.cpp
Comment (pholat) [inline module-apps/apps-common/popups/BluetoothAuthenticatePopup.cpp:84]: all these renames actually add nothing. 
why can't you just change past constexprs:
`style::window::font ....` to the enum you've added?

It will result in much less code change, with literally no difference in these lines. And mass renames result only in less readable PR :(

**Review by pholat — CHANGES_REQUESTED** [review]
First review:
1. please remove changes in configure.sh - these are not needed
2. please add the required documentation for this new feature
3. please revert not needed renames - these pollute the PR and make it unreadable. If you wish to make a mass rename at least add it in separate PR with just that.

File: module-gui/gui/core/FontManager.cpp
Comment (pholat) [inline module-gui/gui/core/FontManager.cpp:17]: so enum was not needed in the first place? :)

File: products/BellHybrid/apps/application-bell-onboarding/windows/OnBoardingOnOffWindow.cpp
Comment (paweljonskim) [reply products/BellHybrid/apps/application-bell-onboarding/windows/OnBoardingOnOffWindow.cpp:31]: I changed that so community fonts would fit. Reverting since we dont wanna make sure they do

File: module-utils/log/api/log/debug.hpp
Comment (pholat) [inline module-utils/log/api/log/debug.hpp:22]: :bow:

File: image/CMakeLists.txt
Comment (paweljonskim) [reply image/CMakeLists.txt:16]: To exclude LUTS.bin from linux builds

File: module-gui/gui/core/FontManager.cpp
Comment (paweljonskim) [reply module-gui/gui/core/FontManager.cpp:80]: handled

File: module-gui/gui/core/FontManager.cpp
Comment (paweljonskim) [reply module-gui/gui/core/FontManager.cpp:109]: added fclose  few lines below

File: module-utils/log/api/log/debug.hpp
Comment (paweljonskim) [reply module-utils/log/api/log/debug.hpp:22]: does this mean an approval?

File: module-gui/gui/widgets/NavBar.cpp
Comment (paweljonskim) [reply module-gui/gui/widgets/NavBar.cpp]: reverted, fixed

File: test/assets/assets_common.json
Comment (paweljonskim) [reply test/assets/assets_common.json]: I autocleaned that via some sed command so didnt notice its empty already.
removed

File: configure.sh
Comment (paweljonskim) [reply configure.sh]: removed all changes on configure.sh layer

File: image/CMakeLists.txt
Comment (paweljonskim) [reply image/CMakeLists.txt]: renamed

File: module-apps/apps-common/popups/BluetoothAuthenticatePopup.cpp
Comment (paweljonskim) [reply module-apps/apps-common/popups/BluetoothAuthenticatePopup.cpp]: reverted all changes to setFont-like methods
also taking in mind:
https://github.com/mudita/MuditaOS/pull/3649#discussion_r829848895
removing Fonts enum and replacing it with string to string mapping (typename to filename ie "medium" -> "gt_pressura_99")

File: module-gui/gui/core/FontManager.cpp
Comment (paweljonskim) [reply module-gui/gui/core/FontManager.cpp]: removed

File: CMakeLists.txt
Comment (pholat) [inline CMakeLists.txt:31]: assets should be an `option` shouldn't these? we actually have quite a few options, please see here:
doc/ProjectConfig.md

File: image/CMakeLists.txt
Comment (pholat) [reply image/CMakeLists.txt:16]: please see that other platform depended things have things like that done by including platform catalog. 

Now - these `JSON` targets are different per option. 

So instead making wonky ifs on strings in here. Why not define `json` target which will be defined based on options combinations to either:
json = json-common + proprietary + rt1051
json = json-common + oss + rt1051
json = json-common + oss
?

File: module-utils/log/api/log/debug.hpp
Comment (pholat) [reply module-utils/log/api/log/debug.hpp:22]: yest it is :)

File: module-gui/test/test-catch-text/CMakeLists.txt
Comment (pholat) [inline module-gui/test/test-catch-text/CMakeLists.txt:14]: I believe you missed `ASSETS_TYPE`.

This is a place where you should rather just move files to ASSET_TYPE dependent catalog and just source them from there. Then you will have one include of each - each ASSET_TYPE dependent one.

**Review by pholat — CHANGES_REQUESTED** [review]
Still seems like needs a bit of work

File: image/CMakeLists.txt
Comment (paweljonskim) [reply image/CMakeLists.txt:16]: well isnt like we are doing this already by defining target `assets`? All it does is to combine different jsons (like you mentioned into one target. How it would be different from the `assets` target?

Comment (pholat) [reply image/CMakeLists.txt:16]: actually `add_assets_target(` calls our cmake function which does some magic, which creates targets and does some more magic. The easiest explanation I can find is:

In my opinion, it would be better to have the approach:
`if -> call to function ` 
I see it as the classic approach and it would mean to include target-dependent things, not add string comparisons in the function call.

Right now you are doing:
```
foo ( if ( something == propritetary) add_something (else if something == community) add_something_else && if (project_target == rt1051 ) add_rt1051)
```
where I think would make a bit more sense to:
```
list(actuall_dependencies)
# somewhere: ...

# 1: have a safe if's
if ( something == proprietary ) list(acctuall_dependencies append propritetary-json)
if ( something ==community) list(actuall_dependencies append community-json)

# 2: actually split PROJECT_TARGET dependent elements
if ( something != proprietary && something != community ) message(ERROR "This is fail safe: wrong $ASSETS_TARGET only proprietary and community available)
include ( ${PROJECT_TARGET}/assets)

# ... 

# then pass it to:
add_assets_target( .... DEPENDS actuall_dependencies)
```

In the end: yes. It's only about organizing the code here.  At best I would see: the removal hack of development assets too as a target, not wonky function call in `add_assets_target` and I know we have to live with what we have. I would just love if we could write it a bit more readable at least :) 

If you disagree. Then it's fine with me. I'm no expert in cmake and I do not feel this is the most important part of my review anyway :)

**Comment by pholat** [comment]
What still needs some work (update from: https://github.com/mudita/MuditaOS/pull/3649#pullrequestreview-914131602 ) 

- [ ] Write documentation for this feature
- [ ] double check ASSETS_TARGET ( I believe in some cases it was not renamed properly and it shouldn't work) 
- [ ] add at least in one place check if ASSETS_TARGET has a proper string value ( I still think that set of OPTIONS would be better, but it doesn't matter)

What's **not required** for me to unblock but **would be highly appreciated**:
- [ ] Try to write cmakes in a fail-fast approach.

Unfortunately, cmake is one of these wonky languages where all variables are just strings and no variable is just an empty string on its own. This promotes situations where no variable - which normally we would treat as an error, would be treated as valid textual input.

File: doc/fonts.md
Comment (pholat) [inline doc/fonts.md:36]: consider just linking to existing example in code :bow: and just pasting here one or two exemplary lines

File: doc/fonts.md
Comment (pholat) [inline doc/fonts.md:48]:
```
FontManager loads all fonts based on `fontmap.json`. These can be later accessed in GUI methods, specifically via: 
```

File: doc/fonts.md
Comment (pholat) [inline doc/fonts.md:46]: Consider adding some more info that:

- MuditaOS uses its own binary format
- That means that all fonts to be used in format have to be first converted to it
- Would be nice to link to font converter (which is an opensource project fork adapted to export our fonts @mjan-dev can we open it ?) 

This is done to:
1. lessen the workload on GPUless processor
2. lessen the size required for fonts to be stored on disc - and loaded on RAM

File: doc/i18n.md
Comment (pholat) [inline doc/i18n.md:103]: 1. There is no GT pressura on harmony? :o 
2.  i18n is just a product-agnostic library - why add it here? add it to the product specification if required ( Pure Phone features are linked from main readme )

File: doc/quickstart.md
Comment (pholat) [inline doc/quickstart.md:107]: users still have to configure the token - not configuration of the token is not mentioned explicitly. 

Can you ignore adding the token when you download nonproprietary assets? No, you can not. Therefore it has to be mentioned here.

File: module-gui/gui/core/FontManager.cpp
Comment (pholat) [inline module-gui/gui/core/FontManager.cpp:102]: as we change Font Manager - can we add some UT here? i.e. if all loaded are loaded, what happens when json states that there is font but there is no such font on disc etc
i.e. here: module-gui/test/test-catch-text/test-gui-Font.cpp

File: module-gui/gui/core/FontManager.cpp
Comment (pholat) [inline module-gui/gui/core/FontManager.cpp:209]: why not: https://en.cppreference.com/w/cpp/container/map/find ?

**Review by pholat — APPROVED** [review]
:+1: Great work :D we accidentally killed a few birds with one stone here.
- we finally have open assets for the OSS community :bow: 
but as a plus:
- we can simply change fonts for different styles & fonts for tests with simple config ( there is actually task for it on roadmap )

File: doc/fonts.md
Comment (mjan-dev) [reply doc/fonts.md:38]: @pholat I believe, we can open it. Lets talk about opening this one and updater at once.

**Pull request merged**
